internal/memory/backend.go

```diff
@@ -0,0 +1,189 @@
+}
+
+type backendStore struct {
+	distClient         backendv1beta1connect.DistributionServiceClient
+	store              *store
+	repoKey            *backendv1beta1.RepositoryKey
+	apikey, sessionKey string
+	wg                 sync.WaitGroup
+	updateInterval     time.Duration
+	cancel             context.CancelFunc
+}
+
+// Close implements Store.
+func (b *backendStore) Close(ctx context.Context) error {
+	// cancel any ongoing background loops
+	b.cancel()
+	// wait for background work to complete
+	b.wg.Wait()
+	_, err := b.distClient.DeregisterClient(ctx, connect.NewRequest(&backendv1beta1.DeregisterClientRequest{
+		SessionKey: b.sessionKey,
+	}))
+	return err
+}
+
+// Evaluate implements Store.
+func (b *backendStore) Evaluate(key string, namespace string, lc map[string]interface{}, dest protoreflect.ProtoMessage) error {
+	return b.store.evaluateType(key, namespace, lc, dest)
+}
+
+func (b *backendStore) registerWithBackoff(ctx context.Context) error {
+	req := connect.NewRequest(&backendv1beta1.RegisterClientRequest{
+		RepoKey:       b.repoKey,
+		NamespaceList: []string{}, // register all namespaces
+	})
+	req.Header().Set(lekkoAPIKeyHeader, b.apikey)
+	var resp *connect.Response[backendv1beta1.RegisterClientResponse]
+	var err error
+	op := func() error {
+		resp, err = b.distClient.RegisterClient(ctx, req)
+		return err
+	}
+
+	exp := backoff.NewExponentialBackOff()
+	exp.MaxElapsedTime = 5 * time.Second
+	if err := backoff.Retry(op, exp); err != nil {
+		return err
+	}
+	if resp != nil {
+		b.sessionKey = resp.Msg.GetSessionKey()
+	}
+	return nil
+}
+
+func (b *backendStore) updateStoreWithBackoff(ctx context.Context) (bool, error) {
+	req := connect.NewRequest(&backendv1beta1.GetRepositoryContentsRequest{
+		RepoKey:    b.repoKey,
+		SessionKey: b.sessionKey,
+	})
+	req.Header().Set(lekkoAPIKeyHeader, b.apikey)
+	var contents *backendv1beta1.GetRepositoryContentsResponse
+	op := func() error {
+		resp, err := b.distClient.GetRepositoryContents(ctx, req)
+		if err != nil {
+			return errors.Wrap(err, "get repository contents from backend")
+		}
+		contents = resp.Msg
+		return nil
+	}
+	exp := backoff.NewExponentialBackOff()
+	exp.MaxElapsedTime = 10 * time.Second
+	if err := backoff.Retry(op, exp); err != nil {
+		return false, err
+	}
+	updated := b.store.update(contents)
+	return updated, nil
+}
+
+func (b *backendStore) shouldUpdateStore(ctx context.Context) (bool, error) {
+	req := connect.NewRequest(&backendv1beta1.GetRepositoryVersionRequest{
+		RepoKey:    b.repoKey,
+		SessionKey: b.sessionKey,
+	})
+	req.Header().Set(lekkoAPIKeyHeader, b.apikey)
+	resp, err := b.distClient.GetRepositoryVersion(ctx, req)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get repository version")
+	}
+	should := b.store.shouldUpdate(resp.Msg.GetCommitSha())
+	return should, nil
+}
+
+func (b *backendStore) loop(ctx context.Context) {
+	b.wg.Add(1)
+	go func() {
+		defer b.wg.Done()
+		after := time.After(b.updateInterval)
```

Comment:
can you use `time.NewTicker` here?

Reply:
updated. i think this was a newer addition to the time package. one nice property of `time.After` is that you can use dynamic configuration to change the interval at runtime if you want. but since we're not planning to do that here, ticker is probably more efficient since we create just one ticker channel throughout the life of the program.